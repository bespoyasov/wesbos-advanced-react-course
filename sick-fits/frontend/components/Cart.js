import React from 'react'
import {Query, Mutation} from 'react-apollo'
import {adopt} from 'react-adopt'
import gql from 'graphql-tag'

import User from './User'
import CartItem from './CartItem'
import TakeMyMoney from './TakeMyMoney'
import calcTotalPrice from '../lib/calcTotalPrice'
import formatMoney from '../lib/formatMoney'

import CartStyles from './styles/CartStyles'
import Supreme from './styles/Supreme'
import CloseButton from './styles/CloseButton'
import SickButton from './styles/SickButton'

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`


const Composed = adopt({
  // render used to remove errors in console
  user: ({render}) => 
    <User>{render}</User>,
  toggleCart: ({render}) => 
    <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
  localState: ({render}) => 
    <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
})

const Cart = props => (
  <Composed>
    {({ user, toggleCart, localState }) => {
      const {self} = user.data
      const {data} = localState
      if (!self) return null

      return (
        <CartStyles open={data.cartOpen}>
          <header>
            <CloseButton 
              onClick={toggleCart}
              title='close'>&times;</CloseButton>
            <Supreme>{self.name}'s cart</Supreme>
            <p>You have {self.cart.length} item{self.cart.length === 1 ? '' : 's'} in your cart</p>
          </header>

          <ul>
            {self.cart.map(cartItem => 
              <CartItem 
                cartItem={cartItem}
                key={cartItem.id} />)}
          </ul>
  
          <footer>
            <p>{formatMoney(calcTotalPrice(self.cart))}</p>
            {!!self.cart.length && (
              <TakeMyMoney>
                <SickButton>Checkout</SickButton>
              </TakeMyMoney>)}
          </footer>
        </CartStyles>
      )
    }}
  </Composed>
)

export default Cart