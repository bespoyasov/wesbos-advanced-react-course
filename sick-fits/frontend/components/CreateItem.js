import React, {Component} from 'react'
import {Mutation} from 'react-apollo'
import gql from 'graphql-tag'
import Router from 'next/router'

import Form from './styles/Form'
import Error from './ErrorMessage'
import format from '../lib/formatMoney'

import {
  imageUploadPreset as preset,
  imageUploadEndpoint as endpoint,
} from '../config'


export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: 'Some title',
    description: 'Some description',
    image: '',
    largeImage: '',
    price: 1500,
  }

  handleChange = e => {
    const {name, type, value} = e.target
    const val = type === 'number' ? parseFloat(value) : value
    return this.setState(state => ({ [name]: val }))
  }

  uploadFile = async e => {
    const files = e.target.files
    const data = new FormData()
    data.append('file', files[0])
    data.append('upload_preset', preset)

    const response = await fetch(endpoint, { method: 'POST', body: data })
    const file = await response.json()

    return this.setState(state => ({ 
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    }))
  }

  render() {
    const {title, price, description, image} = this.state

    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, {loading, error}) => (
          <Form onSubmit={async e => {
            e.preventDefault() 
            const response = await createItem()
            Router.push({
              pathname: '/item',
              query: {id: response.data.createItem.id}
            })
          }}>
            <Error error={error} />
            
            <fieldset disabled={loading} aria-busy={loading}>
              <label>
                Upload an image
                <input 
                  type='file' 
                  name='file' 
                  placeholder='Upload an image' 
                  onChange={this.uploadFile}
                  required />

                {!!image 
                  && <img 
                      alt='Upload preview'
                      width='200'
                      src={image} />}
              </label>

              <label>
                Title
                <input 
                  type='text' 
                  name='title' 
                  placeholder='Title' 
                  value={title}
                  onChange={this.handleChange}
                  required />
              </label>

              <label>
                Price
                <input 
                  type='number' 
                  name='price' 
                  placeholder='Price' 
                  value={price}
                  onChange={this.handleChange}
                  required />
              </label>

              <label>
                Description
                <textarea 
                  name='description' 
                  placeholder='Description' 
                  value={description}
                  onChange={this.handleChange}
                  required />
              </label>

              <button type='submit'>Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem