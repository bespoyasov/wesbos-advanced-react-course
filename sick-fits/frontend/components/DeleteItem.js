import React, {Component} from 'react'
import {Mutation} from 'react-apollo'
import gql from 'graphql-tag'
import {ALL_ITEMS_QUERY} from './Items'


export const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`

class DeleteItem extends Component {
  // manually update cache on the client
  update = (cache, payload) => {
    const data = cache.readQuery({query: ALL_ITEMS_QUERY})
    data.items = data.items.filter(item => 
      item.id !== payload.data.deleteItem.id)

    cache.writeQuery({query: ALL_ITEMS_QUERY, data})
  }

  render() {
    const {children, id} = this.props
    
    return (
      <Mutation 
        mutation={DELETE_ITEM_MUTATION} 
        update={this.update}
        variables={{id}}>
        {(deleteItem, {error}) => {
          return <button onClick={() => {
            if (confirm('Are you sure?')) {
              deleteItem().catch(err => alert(err.message))
            }
          }}>{children}</button>
        }}
      </Mutation>
    )
  }
}

export default DeleteItem