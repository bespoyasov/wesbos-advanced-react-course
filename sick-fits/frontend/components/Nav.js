import Link from 'next/link'

const Nav = () => {
  return (
    <nav>
      <ul>
        <li><Link href='/sell'><a>Sell page</a></Link></li>
        <li><Link href='/'><a>Go home!</a></Link></li>
      </ul>
    </nav>
  )
}

export default Nav