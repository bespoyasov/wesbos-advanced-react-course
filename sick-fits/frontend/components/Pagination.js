import React from 'react'
import PaginationStyles from './styles/PaginationStyles'
import gql from 'graphql-tag'
import {Query} from 'react-apollo'
import Head from 'next/head'
import Link from 'next/link'

import Error from './ErrorMessage'
import {prePage, perPage} from '../config'

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`

const Pagination = (props) => {
  return (
    <Query query={PAGINATION_QUERY}>
      {({data, loading, error}) => {
        if (loading) return <p>Loading...</p>
        if (error) return <Error error={error} />
        
        const {page} = props
        const count = data.itemsConnection.aggregate.count
        const pages = Math.ceil(count / perPage)
        const label = `Page ${page} of ${pages}`
        
        return (
          <PaginationStyles>
            <Head><title>Sick Fits! | {label}</title></Head>
            <Link prefetch href={{
              pathname: 'items', 
              query: {page: page - 1}}}>
              <a className='prev' aria-disabled={page <= 1}>&larr; Prev</a>
            </Link>
            
            <p>{label}</p>
            <p>{count} items total</p>

            <Link prefetch href={{
              pathname: 'items', 
              query: {page: page + 1}}}>
              <a className='next' aria-disabled={page >= pages}>Next &rarr;</a>
            </Link>
          </PaginationStyles>
        )
      }}
    </Query>
  )
}

export default Pagination