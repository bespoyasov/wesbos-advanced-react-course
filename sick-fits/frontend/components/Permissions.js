import {Query} from 'react-apollo'
import gql from 'graphql-tag'
import Error from './ErrorMessage'

const ALL_USERS_QUERY = gql`
  query {
    users {
      id
      name
      email
      permissions
    }
  }
`

const Permissions = props => (
  <Query query={ALL_USERS_QUERY}>
    {({data, loading, error}) => (
      <>
        <Error error={error} />
        <p>asdasdasdas</p>
      </>
    )}
  </Query>
)

export default Permissions