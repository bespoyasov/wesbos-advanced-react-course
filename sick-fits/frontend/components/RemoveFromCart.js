import React, {Component} from 'react'
import {Mutation} from 'react-apollo'
import styled from 'styled-components'
import gql from 'graphql-tag'
import PropTypes from 'prop-types'
import {CURRENT_USER_QUERY} from './User'


export const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`


const BigButton = styled.button`
  font-size: 3rem;
  border: 0;
  background: none;
  cursor: pointer;
  &:hover {
    color: ${props => props.theme.red};
  }
`


class RemoveFromCart extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  }

  // gets called as soon as we get a response from the server 
  // after a mutation has been performed
  update = (cache, payload) => {
    // read the cache
    const data = cache.readQuery({query: CURRENT_USER_QUERY})

    // remove that item from the cart
    const cartItemId = payload.data.removeFromCart.id
    data.self.cart = data.self.cart.filter(cartItem => 
      cartItem.id !== cartItemId)
    
    // write it back to the cache
    cache.writeQuery({
      query: CURRENT_USER_QUERY,
      data
    })
  }

  render() {
    const {id} = this.props

    return (
      <Mutation 
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{id}}
        update={this.update}
        optimisticResponse={{
          __typename: 'Mutation',
          removeFromCart: {
            __typename: 'CartItem',
            id
          }
        }}>
        {(removeFromCart, {loading, error}) => 
          <BigButton 
            disabled={loading}
            title='Delete item'
            onClick={() => {
              removeFromCart().catch(err => alert(err.message))
            }}>
            &times;
          </BigButton>}
      </Mutation>
    )
  }
}

export default RemoveFromCart