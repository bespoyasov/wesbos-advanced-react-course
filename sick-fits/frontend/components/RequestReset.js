import React, {Component} from 'react'
import {Mutation} from 'react-apollo'
import gql from 'graphql-tag'

import Form from './styles/Form'
import Error from './ErrorMessage'


const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`


const initialState = {
  email: '',
}


class RequestReset extends Component {
  state = {...initialState}

  updateState = (e) => {
    const {name, value} = e.target
    this.setState(state => ({ [name]: value }))
  }

  render() {
    const {email} = this.state

    return (
      <Mutation 
        mutation={REQUEST_RESET_MUTATION} 
        variables={this.state}>
        {(requestReset, {error, loading, called}) => {
          return (
            <Form method='post' onSubmit={async e => {
              e.preventDefault()
              await requestReset()
            }}>
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Reset a password</h2>
                <Error error={error} />

                {!error && !loading && called && (
                  <p>Success! Check your email</p>
                )}

                <label>
                  Email
                  <input 
                    type='email' 
                    name='email'
                    placehodler='Email'
                    value={email} 
                    onChange={this.updateState} />
                </label>
      
                <button type='submit'>Submit!</button>
              </fieldset>
            </Form>
          )
        }}
      </Mutation>
    )
  }
}

export default RequestReset