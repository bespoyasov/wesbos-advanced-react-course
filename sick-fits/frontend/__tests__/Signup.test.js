import {mount} from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import {MockedProvider} from 'react-apollo/test-utils'
import {ApolloConsumer} from 'react-apollo'

import Signup, {SIGNUP_MUTATION} from '../components/Signup'
import {CURRENT_USER_QUERY} from '../components/User'
import {fakeUser} from '../lib/testUtils'


function type(wrapper, name, value) {
  wrapper
    .find(`input[name="${name}"]`)
    .simulate('change', {
      target: {name, value}
    })
}

const self = fakeUser()

const mocks = [{
  // signup mock mutation
  request: {
    query: SIGNUP_MUTATION,
    variables: {
      email: self.email,
      name: self.name,
      password: 'test'
    }
  },
  result: {
    data: {
      signup: {
        __typename: 'User',
        id: 'abs123',
        email: self.email,
        name: self.name,
      }
    }
  }
}, {
  // current user query mock
  request: {
    query: CURRENT_USER_QUERY,
  },
  result: {
    data: {self}
  }
}]


describe('<Signup />', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <Signup />
      </MockedProvider>
    )

    expect(toJSON(wrapper.find('form'))).toMatchSnapshot()
  })

  it('calls the mutation', async () => {
    let apolloClient

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client
            return <Signup />
          }}
        </ApolloConsumer>
      </MockedProvider>
    )

    await wait()
    wrapper.update()

    type(wrapper, 'name', self.name)
    type(wrapper, 'email', self.email)
    type(wrapper, 'password', 'test')
    wrapper.update()
    wrapper.find('form').simulate('submit')
    
    await wait()
    
    // query the user out of the apollo client
    const user = await apolloClient.query({query: CURRENT_USER_QUERY})
    expect(user.data.self).toMatchObject(self)
  })
})