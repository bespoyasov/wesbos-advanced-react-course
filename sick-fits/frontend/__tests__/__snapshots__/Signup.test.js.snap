// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Signup for an account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label>
      Email
      <input
        name="email"
        onChange={[Function]}
        placehodler="Email"
        type="email"
        value=""
      />
    </label>
    <label>
      Name
      <input
        name="name"
        onChange={[Function]}
        placehodler="Name"
        type="text"
        value=""
      />
    </label>
    <label>
      Password
      <input
        name="password"
        onChange={[Function]}
        placehodler="Password"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Signup!
    </button>
  </fieldset>
</form>
`;
